### ISSUE TITLE
Datastore: receiving duplicate entities during query iteration

### ISSUE THREAD

**Issue by berlincode** [GoogleCloudPlatform/google-cloud-python#4264]
Hi.

I'm using google-cloud-datastore==1.3.0 from pypi.

When iterating over a query iterator the google-cloud-datastore sometimes returns the same entitiy twice.

I've had a look at the source and it seems that the last entity of a fetched page is the same as the first entity of the next page.

Iterating of pages and printing each enitity id makes the problem obvious.
```python
query_iter = qry.fetch()
pages = query_iter.pages
while True:
    page = next(pages)
    if not page:
        return

    for ent in page:
        print(ent.key.id)

    print("-- new page --")
```

The output looks like this:
```
6438744212111360
5774037759295488
5226622299930624
6390563805855744
5537056630505472
5007404484788224
5289011263307776
5805078561685504
-- new page --
5805078561685504
4785039263924224
4655690686660608
5253444773347328
5481464721309696
```

As you can see the entity 5805078561685504 is returned twice.

Dumping the protocol buffer shows that the 'cursor'-value of the last entity and is identical to the 'end_cursor'-value. Here are the last lines of the dump:```

```
      }
    }
  }
  cursor: "\n\346\002\n\260\002\n\003ka2\022\250\002\032\245\002..."
  version: 1507954620963000
}
end_cursor: "\n\346\002\n\260\002\n\003ka2\022\250\002\032\245\002..."
more_results: NOT_FINISHED
```

Maybe you can help.

**Comment by dhermes** [comment]
Thanks for reporting! We'll look into it. (Also, I'm happy to see you've been able to use the `pages` feature of the iterator!)

**Comment by sagarrakshe** [comment]
Hey @berlincode you mentioned it happens _sometimes_. I tried with the same version of datastore, I wasn't able to reproduce the issue. It returned me all unique entities, no duplicates. Any specific steps or requirements to reproduce this issue?

**Comment by berlincode** [comment]
Withe the example above it is completely reproducible each time.

I've had the "feeling", that other queries may have the same problem.

Maybe it's also relevant to say that the project was started as an app engine standard project and was later migrated to app engine flexible.

If there is someone with access to my datastore: my project_id is "robotdetect-com". The code is run under python 3.6.2.

Here is a complete example (including the query) to reproduce the problem:
```
ds_client = <setup datastore client here>


qry = ds_client.query(kind="ResultArray", order=["ka2"])
qry.add_filter("ka2", ">=", b'  Root-         adspert-           a0015-0001507939200000{')

query_iter = qry.fetch(start_cursor=None)

pages = query_iter.pages
page = next(pages)

for ent in page:
    print(ent.key.id)

cursor = query_iter.next_page_token

print("--")

query_iter = qry.fetch(start_cursor=cursor)

pages = query_iter.pages
page = next(pages)

for ent in page:
    print(ent.key.id)
```

**Comment by tseaver** [comment]
@berlincode I just tried to build a system test which reproduces your case (see #4905).  That test, which *does* have multiple pages in the response, passes.

Please reopen if you can provide a complete case which reproduces the issue (i.e., it actually populates entities and then queries them with the duplicate).

**Comment by berlincode** [comment]
Hello. It's sad that the ticket is closed. The problem is still exists and is reproduceable on my datastore (project_id is "robotdetect-com") with the code shown above. Can please someone from google with access to the datastore can confirm that the problem exists?

It took quite some time to track this down.

I can only guess that this problem has somehow to do with the very long keys that are used. The problem does not occur on every paged query, but only sometimes. Additionally I've got the feeling that this might depend on how the paged data is served.

I'm not sure if i can create a complete test.

**Comment by tseaver** [comment]
@berlincode the example I added in PR #4905 uses UUIDs as keys.  Can you look at how it is put together and see if something in your own code jumps out as different?

**Comment by berlincode** [comment]
First I want to thank you that you wrote a test case.

I've invested some more time and maybe the results may get us a bit further:

1) First I had a look at your test code. I've modified it a bit to match my query and to work on ids:

```python

page_query = ds_client.query(kind="ResultArray", order=["ka2"])
page_query.add_filter("ka2", ">=", b'  Root-         adspert-           a0015-0001507939200000{')

iterator = page_query.fetch()

seen = set()
page_count = 0
for page in iterator.pages:
    page_count += 1

    for entity in page:
        #uuid_str = entity.key.name
        uuid_str = entity.key.id # <- changed test to use id
        if uuid_str in seen:
            raise Exception("found duplicate: %s in page nr %d" % (str(uuid_str), page_count))
        seen.add(uuid_str)

```
Then i get following Exception "Exception: found duplicate: 5805078561685504 in page nr 2"

So the problem still exists atl the same entity as shown at top. I've run this a couple of times and his seems to be reproducible always at that entity.

2) If I add a "page_query.keys_only()" to the query I'm receiving (of course) more entities (300) within the first page and the exception is thrown at another entity:
"Exception: found duplicate: 6610136358453248 in page nr 2"

3) I've failed again to create a complete test case with synthetic entities.

How can I help to resolve this issue?

**Comment by tseaver** [comment]
OK, I'll reopen.

Is there any possibility that your entities have a collision on the IDs?

**Comment by berlincode** [comment]
I'm not sure how to test for collisions on IDs.

But I think a collision on IDs is rather unlikely: The problem is only visible (if at all) at the boundary of the paged query. Additionally (as I wrote above) if I added "page_query.keys_only()" the duplicate ID changed to 6610136358453248  (and is again on a page boundary) while the ID 5805078561685504 (which was repeated ID without the keys_only) then is only returned once.

Maybe the problem is the result of lower layer e.g. protobuf or even deeper.

Is there a simple way to create a dump (e.g. protorpc) which may be helpful?

p.s. Thanks for reopening this issue.

**Comment by jonparrott** [comment]
Are you modifying the entities at all while querying?

**Comment by berlincode** [comment]
Just to be sure that this is not the source of the problem I just ran a paged query from the latest entities to older ones. The first duplicate is just ~2h old and for this entity I'm sure that it was only written once and never updated.

In general, yes, if the entity model has change (e.g. new properties) or I modified datastore indexes I sometimes iterate over entities and do an update. Typically this is very rare and typically I only update the entities of the last ~3 days.
